ENH: Add vectorized scalar minimization bracket finder

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
I don't think there's an issue for this.

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR adds a private vectorized `scalar_minimization` bracket finder, similar to the vectorized root bracket finder from gh-18348. The intent of this function is the same as [scipy.optimize.bracket](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.bracket.html). A bracket for the minima of a unimodal univariate function $f$ is a triple of points $x_l$, $x_m$, $x_r$, with $x_l < x_m < x_r$, such that

$$f(x_l) \geq f(x_m),\text{  }f(x_r) \geq f(x_m)$$

where at least one of these two inequalities is strict. Unlike `scipy.optimize.bracket`, this function is able to accept upper
and lower bounds on the endpoints of brackets. The algorithm was suggested to me by @mdhaber, who convinced me it is a sound approach. Quoting from the Notes section of the Docstring

    Given an initial trio of points `xl`, `xm`, `xr`, the algorithm checks if
    these points already give a valid bracket. If not, a new endpoint, `w` is
    chosen in the "downhill" direction, `xm` becomes the new opposite endpoint,
    and either `xl` or `xr` becomes the new middle point, depending on which
    direction is downhill. The algorithm repeats from here.

    The new endpoint `w` is chosen differently depending on whether or not a
    boundary `min` or `max` has been set in the downhill direction. Without
    loss of generality, suppose the downhill direction is to the right, so that
    ``f(xl) > f(xm) > f(xr)``. If there is no boundary to the right, then `w`
    is chosen to be ``xr + phi * (xr - xm)`` where `phi` is the golden ratio;
    so that step sizes increase in geometric proportion, with each step size
    equal to the sum of the two previous step sizes. If there is a boundary,
    `max` in this case, then `w` is chosen to be ``xr + (max - xr)/phi``, so
    that step sizes decrease in geometric proportion, slowing to a stop at
    `max`. This cautious approach ensures that a minima near but distinct from
    the boundary isn't missed.

Given a 
#### Additional information
<!--Any additional information you think is important.-->
This has been placed in `scipy.optimize._zeros_py.py` to avoid a circular import, since that is where most of the machinery used here lives. A more rightful home would be `scipy.optimize._optimize.py` where `scipy.optimize.bracket` lives. Since this is private, I don't think this is important just yet, but we should keep it in mind for the future. @mdhaber,  I think I finally get the details of how this machinery works.

mdhaber reviewed (commented):
I'm flying now, so I'd like to play with it a bit when I get to my desktop, but it looks good!

> This has been placed in `scipy.optimize._zeros_py.py` to avoid a circular import, since that is where most of the machinery used here lives.

Shall I move the machinery before we merge this?

> I think I finally get the details of how this machinery works.

Cool! In that case, would you review gh-19545? It relaxes the requirement that the shape of the function output must be the shape of the input. This is much more convenient in some cases (especially `_tanhsinh`, in which case users want to express callables as vector-valued).

mdhaber commented on scipy/optimize/_zeros_py.py:
```diff
@@ -2866,3 +2866,240 @@ def _differentiate_weights(work, n):
+    if not np.issubdtype(a.dtype, np.number) or np.iscomplex(a).any():
+        raise ValueError('`a` must be numeric and real.')
+
+    b = a + 1 if b is None else b
```
This could be beyond max.

mdhaber commented:
Here's a pretty useful smoke test. I adjusted `b = a + 0.1` to reduce the number of invalid input errors.

<details>

```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(2549824598234528)

from scipy.optimize._zeros_py import _bracket_minima
from scipy.stats._distr_params import distcont
distcont = dict(distcont)

n_trials = 0
n_success = 0

for distname, params in distcont.items():

    family = getattr(stats, distname)
    dist = family(*params)

    def f(x):
        return -dist.pdf(x)

    try:
        res = _bracket_minima(f, a=dist.rvs(size=10),
                              min=dist.support()[0], max=dist.support()[1])
        print(distname, np.all(res.success))
        n_trials += 10
        n_success += np.sum(res.success)
    except Exception as e:
        print(distname, str(e))
```

</details>

Obviously there's no reason that all of these should achieve success, but it does pretty well! 97% success rate adjusting only for the cases which got an invalid input error.

mdhaber commented on scipy/optimize/_zeros_py.py:
```diff
@@ -2866,3 +2866,240 @@ def _differentiate_weights(work, n):
+    # We only need the boundary in the direction we're traveling.
+    limit = np.where(comp, min, max)
+
+    golden_ratio = 1.618033988749895
```
Despite this, the code seems to respect `dtype`, probably because `step` has the `dtype` of `a`. But it might be better to explicitly cast to `dtype` to avoid conversion each iteration. I don't think there are any other constants to worry about.

steppi commented:
Thanks @mdhaber. Good points all. I have a clear idea of how to proceed for everything except how to handle the arguments for left, middle and right initial points. I don't have a strong preference yet, but will think about it some more and then we can discuss.

steppi commented:
@mdhaber. I think I've made all of the code changes asked for. I changed the signature to include the midpoint as a positional arg, and made the left and right endpoints optional and keyword only. I wrote a scheme for the defaults to ensure the initial bracket won't go beyond a bound when the relevant endpoint isn't specified. I also changed how the updates work to match `bracket_root`, and made `factor` a parameter which can be an array.

I haven't updated the tests yet. There's still a question of standardizing variable names for local variables in the functions, and for the inputs. As a proposal, I've changed `min` and `max` to `xmin` and `xmax`, because I don't want to encourage overwriting the builtin `min` and `max` functions. We know better to avoid trouble, but a user might write something like

```python
...
max = [1.0, 2.0, 3.0]
result = bracket_minimum(f, xm, max=max, args=args)
```

I also changed `a`, `b`, `c`, to `xl`, `xr`, `xm`, since I think `a`, `b`, `c` are not as clear for left, right, and middle points of bracket.

and run into problems

lucascolley commented:
docs build will be fixed upon merging main

mdhaber commented on scipy/optimize/_zeros_py.py:
```diff
@@ -2866,3 +2866,240 @@ def _differentiate_weights(work, n):
+    `scipy.optimize.bracket`, users may specify minimum and maximum endpoints
+    for the desired bracket.
+
+    Given an initial trio of points `xl`, `xm`, `xr`, the algorithm checks if
```
Well, now, a lot of these should technically be single backticks because they are parameters.

mdhaber commented on scipy/optimize/_zeros_py.py:
```diff
@@ -2866,3 +2866,261 @@ def _differentiate_weights(work, n):
+                      ('fr', 'fr')]
+
+    def pre_func_eval(work):
+        work.step[~work.limited] *= work.factor[~work.limited]
+        work.step[work.limited] /= work.factor[work.limited]
+        x = np.empty_like(work.xr)
+        x[~work.limited] = work.xr[~work.limited] + work.step[~work.limited]
+        x[work.limited] = work.limit[work.limited] - work.step[work.limited]
```
Thanks, this is more similar to the `_bracket_root` logict now. It makes me realize, though, that I think we could redefine factor at the top like
```python3
work.factor[work.limited] = 1/work.factor[work.limited]
```

and then we don't have do fancy indexing every iteration.

```python3
work.step *= work.factor
```

steppi replied:
I've changed it for `bracket_minimum`. I think a separate PR for the other function would be better. I had to change

```python
    factor = np.broadcast_to(factor, shape).astype(dtype, copy=False).ravel()
```

to 

```python
    factor = np.broadcast_to(factor, shape).copy().astype(dtype, copy=False).ravel()
```

because `np.broadcast_to` produces a read only view. I guess the `copy=False` might not be needed anymore, but I left it in.

steppi commented:
OK. I've tested this thoroughly and actually uncovered a couple bugs. 1) dtypes changing to np.float64 in `_bracket_minimum_iv`, 2) not technically a bug, but I was not calculating the new endpoint in the same way as in `_bracket_root`. Also, I've checked that all cases now pass in the smoketests you wrote above. I haven't updated internal variable names to match `_bracket_root`, but maybe that can be left for a followup. I like using fully spelled out names,
because it makes it easier for me to read what's going when my brain starts scrambling things. Now that it seems to be working correctly, I don't mind changing to match the convention from `_bracket_root` though. 

I like `xl`, `xm`, `xr` better than `a`, `b`, `c`, and `xmin` and `xmax` better than `min` and `max` for parameter names. It's probably good to make the two functions consistent though.

steppi commented:
OK. I think I've successfully resolved the merge conflicts. I think the main thing to settle down now is consistent argument names for `_bracket_root` and `_bracket_minimum`.

For the former we currently have `a`, `b`, `min, `max`, but with the convention of the later these would be `xl`, `xr`, `xmin`, `xmax`.

For the later we have `xm`, `xl`, `xr`, `xmin`, `xmax`. Using the convention from the former this would be `c`, `a`, `b`, `min`, `max`.

I prefer the conventions for `bracket_minimum` because

- I find it easy to mix up which points on the interval `a`, `b`, and `c` are supposed to be. Is `b` the midpoint, or `c`? This isn't a problem if we only have `_bracket_root`.
- I don't think we should encourage redefining the built-ins `min` and `max`.

mdhaber commented:
The reason I didn't use `xl` etc... was that these don't have the same properties/meaning as the results, so I didn't think the attributes of the return object should be the same as the names of these inputs. Maybe `xmin`, `xl0`, `xm0`, `xr0`, `xmax`? I don't love the `0`s either because that suggests "guess", and I'm not sure if these should be interpreted as guesses in the same way as `x0` of, say, `newton`. But I agree that we should standardize and the current names aren't ideal.

steppi commented:
> The reason I didn't use `xl` etc... was that these don't have the same properties/meaning as the results, so I didn't think the attributes of the return object should be the same as the names of these inputs. Maybe `xmin`, `xl0`, `xm0`, `xr0`, `xmax`? I don't love the `0`s either because that suggests "guess", and I'm not sure if these should be interpreted as guesses in the same way as `x0` of, say, `newton`. But I agree that we should standardize and the current names aren't ideal.

Oh, good point. I do think of them like initial guesses, so I don't think that aspect is an issue. It kind of bugs me that I can't help reading them as "x lo", "x mo", "x ro". `xl_0`, `xm_0`, xr_0`, is more clear in that regard, but it's annoying to have to type the underscore, and then would we have `x_min`, `x_max` for consistency? I can live with `xl0`, `xm0`, `xr0`.

mdhaber commented:
Would prefer to avoid the underscores, and it's `x0` without the underscore in basically every other interface. So let's go with `xmin`, `xl`, `xm`, `xr`, and `xmax`.

steppi committed:
MAINT: Update bracket finder arg name scheme

steppi commented:
> Would prefer to avoid the underscores, and it's `x0` without the underscore in basically every other interface. So let's go with `xmin`, `xl0`, `xm0`, `xr0`, and `xmax`, at least while they're private.

Done

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+                     minimizer.
+            - ``-2`` : The maximum number of iterations was reached.
+            - ``-3`` : A non-finite value was encountered.
+            - ``-4`` : Iteration was terminated by `callback`.
+            - ``1`` : The algorithm is proceeding normally (in `callback` only).
```
I suppose we can include these while it's private, but technically `callback` is not exposed for now. (But let's add it before this goes public.)
```suggestion
```

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+    -----
+    Similar to `scipy.optimize.bracket`, this function seeks to find real
+    points ``xl < xm < xr`` such that ``f(xl) >= f(xm)`` and ``f(xr) >= f(xm)``,
+     where at least one of the inequalities is strict. Unlike `scipy.optimize.bracket`,
```
```suggestion
    where at least one of the inequalities is strict. Unlike `scipy.optimize.bracket`,
```

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+    `scipy.optimize.bracket`, users may specify minimum and maximum endpoints
+    for the desired bracket.
+
+    Given an initial trio of points `xl = xl0``, ``xm = xm0``, ``xr = xr0``,
```
```suggestion
    Given an initial trio of points ``xl = xl0``, ``xm = xm0``, ``xr = xr0``,
```

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+
+    Given an initial trio of points `xl = xl0``, ``xm = xm0``, ``xr = xr0``,
+    the algorithm checks if these points already give a valid bracket. If not,
+    a new endpoint, `w` is chosen in the "downhill" direction, `xm` becomes the new
```
```suggestion
    a new endpoint, ``w`` is chosen in the "downhill" direction, ``xm`` becomes the new
```
and similar throughout for symbols that don't refer to parameters. I guess the goal is for parameters to link back to their documentation, and all should eventually be rendered in monospaced font.

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+    boundary `xmin` or `xmax` has been set in the downhill direction. Without
+    loss of generality, suppose the downhill direction is to the right, so that
+    ``f(xl) > f(xm) > f(xr)``. If there is no boundary to the right, then `w`
+    is chosen to be ``xr + factor * (xr - xm)`` where `factor` is controlled by
```
I think this means that by default `xr - xm` triples in size each step. I'm not sure if this matters, and I'm not sure what I want, but consider whether this should be `xm + factor * (xr - xm)`? I believe that in `_bracket_root` the total bracket width increases by `factor` each step`. If `(xr - xm) == (xm - xl)`, then I suppose what you have could be interpreted in the same way, but that's not guaranteed, so I'm not sure what is best.

steppi replied:
Huh. It looks like SciPy's existing scalar minimization bracket finder does what you suggest

https://github.com/scipy/scipy/blob/fb4c65fdf89866728a9762b47e1c75c03e0b2777/scipy/optimize/_optimize.py#L2984

(It uses the `a < c < b` notation, but `c` isn't public.) I think I got mixed up because another reference I was checking uses `a < b < c` notation for what is essentially the same basic algorithm. I'll change it to `xm + factor * (xr - xm)`

steppi replied:
I'm not sure which update step is actually better, but it seems to work as is.

mdhaber replied:
Hah ok let's table this. I don't think it matters much; it's just something that I think is worth discussing in real-time at some point.

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+    factor = np.broadcast_to(factor, shape).astype(dtype, copy=True).ravel()
+
+    # To simplify the logic, swap xl and xr if f(xl) < f(xr). We should always be
+    # marching downhill in the direction from a to b.
```
```suggestion
    # marching downhill in the direction from xl to xr.
```

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+        x[~work.limited] = work.xr0[~work.limited] + work.step[~work.limited]
+        x[work.limited] = work.limit[work.limited] - work.step[work.limited]
+        # Since the new bracket endpoint is calculated from an offset with the
+        # limit,it may be the case that the new endpoint equals the old endpoint,
```
```suggestion
        # limit, it may be the case that the new endpoint equals the old endpoint,
```

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+
+    if not np.issubdtype(factor.dtype, np.number) or np.iscomplex(factor).any():
+        raise ValueError('`factor` must be numeric and real.')
+    if not np.all(factor > 1):
+        raise ValueError('All elements of `factor` must be greater than 1.')
```
I seem to remember thinking of cases in which `factor < 1` could be desirable for `_bracket_root`. Anyway, something to think about, but this is the primary use case.

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+    if not np.all(factor > 1):
+        raise ValueError('All elements of `factor` must be greater than 1.')
+
+    # Default choices for xl or xr might have exceeded xmin or xmax. Adjust
+    # to make sure this doesn't happen. We replace with copies because xl, and xr
```
What if they were all provided and this condition occured due to user error?

steppi replied:
Oh good point, it would make sense to only correct the choices of xl and xr if they are default values, not if the user entered them incorrectly.

mdhaber commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+    if not maxiter == maxiter_int or maxiter < 0:
+        raise ValueError(message)
+
+    if not np.all((xmin <= xl0) & (xl0 < xm0) & (xm0 < xr0) & (xr0 <= xmax)):
+        raise ValueError(
+            '`xmin <= xl0 < xm0 < xr0 <= xmax` must be True (elementwise).'
+        )
```
This is fine for now, but I think I did this at some point when I was feeling lazy and set a bad precedent. Bracket calculation should continue but fail for these slices.

steppi replied:
Yeah, that does sound better.

mdhaber reviewed (commented):
I looked at `_bracket.py` and found some minor things. I'll take a look at the tests soon!

mdhaber commented on scipy/optimize/tests/test_bracket.py:
```diff
@@ -286,15 +286,387 @@ def f(x):
+    )
+    @pytest.mark.parametrize("use_xmin", (False, True))
+    @pytest.mark.parametrize("other_side", (False, True))
+    def test_nfev_expected(self, seed, use_xmin, other_side):
```
LMK if you want me to check the details, but like `TestBracketRoot.test_nfev_expected`, this looks like a very strong test that confirms the algorithm is behaving precisely as expected in the scalar case.

mdhaber commented on scipy/optimize/tests/test_bracket.py:
```diff
@@ -286,15 +286,387 @@ def f(x):
+            f, xm0, xl0=xl0, xr0=xr0, xmin=xmin, xmax=xmax, args=(minimum, )
+        )
+        assert np.all(result.success)
+        assert result.xl.dtype == result.xm.dtype == result.xr.dtype == dtype
```
For good measure? It is _possible_ that even if `f` returns the right `dtype` we don't preserve it.
```suggestion
        assert result.xl.dtype == result.xm.dtype == result.xr.dtype == dtype
        assert result.fl.dtype == result.fm.dtype == result.fr.dtype == dtype
```

mdhaber commented on scipy/optimize/tests/test_bracket.py:
```diff
@@ -286,15 +286,387 @@ def f(x):
+        # Initial bracket is valid.
+        f = self.init_f()
+        result = _bracket_minimum(f, -0.2, xl0=-1.0, xr0=1.0, args=(0, 0))
+        assert f.count == 3
```
In `TestBracketRoot.test_special_cases`, I probably should have checked that the `result` has the correct `x`s and `f`s, too. Let's do that for completeness.

It looks like the other special cases, like when the minimum is at or beyond a limit or endpoint, is are tested elsewhere. Can you think of any edge cases that are specific to `_bracket_minimum`, though?

mdhaber commented on scipy/optimize/tests/test_bracket.py:
```diff
@@ -286,15 +286,387 @@ def f(x):
+            (np.array([[0.0], [-0.1]]), np.array([[0.0], [2.5]])),
+        )
+    )
+    def test_vectorized(self, xl0, xm0, xr0, xmin, xmax, args):
```
Most of the other tests followed the pattern of `TestBracketRoot` pretty carefully, and we already decided that those were acceptable, so it's easy to judget that these are acceptable by analogy. This test is quite different - why did you decide to change this one? Is it as strong or stronger in your opinion?

steppi replied:
This is one of the first tests I wrote before looking at the tests for `TestBracketRoot`, and I just neglected to change it after your review feedback. I don't think it's stronger, but it was useful while I was implementing things. I'll change it to match the convention used in `TestBracketRoot`.

mdhaber commented on scipy/optimize/tests/test_bracket.py:
```diff
@@ -286,15 +286,387 @@ def f(x):
+    @pytest.mark.parametrize(
+        "args",
+        (
+            (1.2, 0), (-0.5, 0), (0.1, 0), (0.2, 0), (3.6, 0), (21.4, 0),
+            (121.6, 0), (5764.1, 0), (-6.4, 0), (-12.9, 0), (-146.2, 0)
```
Here and below, could you comment on the magic numbers? Are we testing a lot of conceptually distinct cases, or just testing a small number of conceptually distinct things with many different values?

steppi replied:
I agree the comment are needed here. I'll need to look into this again to understand the motivation behind what I was doing, which really cries out for comments.

mdhaber replied:
Or make the specific number random to emphasize the property you're testing.

mdhaber reviewed (commented):
Looks like the test match those for `TestBracketRoot` for the most part, and I'm not thinking of other tests that would be needed here (that aren't already). The main comments are that the vectorization test looks like it could be stronger and I'd like to know more about the magic numbers in tests from `test_scalar_no_limits` through `test_minima_at_boundary_point`.

steppi commented on scipy/optimize/_bracket.py:
```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+
+            func(x: ndarray, *args) -> ndarray
+
+        where each element of `x` is a finite real and `args` is a tuple,
```
Thanks. Reading through the discussion for your PR should help me internalize the distinction.

steppi committed:
MAINT: Apply suggestions from code review for bracket_minimum

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

Only update xl0/xr0 to feasible values when not supplied

Fix plural when should be singular

Combine xl0, xr0 in docstring

Fix false positive variable renaming

Change vectorization test to match TestBracketRoot

Add comments explaining magic constants in tests

Add another check to initial bracket valid test

Mention that `xl0`, `xr0` must be broadcastable with `xm0`

mdhaber commented:
OK, I think those will be my only comments. When they are resolved, I think it's good to go. 

When it's done, would you do a quick PR to address https://github.com/scipy/scipy/pull/19757#pullrequestreview-1807271162 for `_bracket_root`?

steppi commented:
Thanks @mdhaber! I batched up and committed all of your suggestions and should have time to fix the remaining things on Thursday or Friday.

mdhaber commented:
OK. When that's done, maybe squash commits before and after the name changes so that the name change commit is separate.

steppi commented:
I think I've addressed everything.

- Replaced "minima" with "minimum" in another place in the docstring, because we are sticking with singular.
- Changed vectorization test to match that for `bracket_root`.
- added comments to explain the constants in the tests for `bracket_minimum`
- Make it so `xl0` and `xr0` are updated to be made feasible only if they are taking default values, not if user passed infeasible values.
- Added check that `result.xl`, ... `result.fl` .. are correct in special cases test where initial bracket is valid.
- combined `xl0` and `xr0` in docstring

I've also fixed a place where I accidentally changed `a`, `b` to `xl0`, `xr0`, when these were args for the function `f` that really should have been `a` and `b`. After you think this looks good, I'll do an interactive rebase to group it into 3 commits like you suggested.

mdhaber commented on scipy/optimize/tests/test_bracket.py:
```diff
@@ -569,74 +590,150 @@ def test_scalar_with_limit_right(self, xl0, xm0, xr0, xmax, args):
+        xr0 = rng.random(size=shape)
+        xm0 = xl0 + rng.random(size=shape) * (xr0 - xl0)
+        xmin, xmax = 1e3*xl0, 1e3*xr0
+        if shape:  # make some elements un
```
```suggestion
        if shape:  # make some elements infinite
```
?

steppi replied:
Copy and pasted from the corresponding `bracket_root` without being read.

https://github.com/scipy/scipy/blob/fb4c65fdf89866728a9762b47e1c75c03e0b2777/scipy/optimize/tests/test_bracket.py#L125

mdhaber commented:
Re: `xl0, xr0` "Must be broadcastable with one another." They also must be broadcastable with `xm0`. I've not been sure how to document this sort of thing because in the end, essentially all array arguments, including those in `args`, must be mutually broadcastable with one another. So far, I have been mentioning the broadcasting requirement with all arguments that have already been documented, so this might be like "Must be broadcastable with one another and `xm0`".

mdhaber commented:
OK, after considering these little tweaks, please feel free to rebase.

steppi commented:
Well, that attempt at git surgery was less than successful. Sorry everyone who got notified.

steppi commented:
@mdhaber, I'm just going to close this and make a new PR with the three commits.

steppi committed:
ENH: Add vectorized scalar minimization bracket finder

steppi commented:
Here's the backup of my branch so you can check the diff https://github.com/steppi/scipy/tree/scalar-minimize-bracket-finder-backup
